store/sqlstore/supplier.go

```diff
@@ -1319,3 +1305,59 @@ func convertMySQLFullTextColumnsToPostgres(columnNames string) string {
 
 	return concatenatedColumnNames
 }
+
+func checkDatabaseConnection(settings *model.SqlSettings, dataSource, connType string) *dbsql.DB {
+	db, err := dbsql.Open(*settings.DriverName, dataSource)
+	if err != nil {
+		mlog.Critical("Failed to open SQL connection to err.", mlog.Err(err))
+		time.Sleep(time.Second)
+		os.Exit(EXIT_DB_OPEN)
+	}
+
+	for i := 0; i < DB_PING_ATTEMPTS; i++ {
+		mlog.Info("Pinging SQL", mlog.String("database", connType))
+		ctx, cancel := context.WithTimeout(context.Background(), DB_PING_TIMEOUT_SECS*time.Second)
+		defer cancel()
+		err = db.PingContext(ctx)
+		if err == nil {
+			break
+		} else {
+			if i == DB_PING_ATTEMPTS-1 {
+				mlog.Critical("Failed to ping DB, server will exit.", mlog.Err(err))
+				time.Sleep(time.Second)
+				os.Exit(EXIT_PING)
+			} else {
+				mlog.Error("Failed to ping DB", mlog.Err(err), mlog.Int("retrying in seconds", DB_PING_TIMEOUT_SECS))
+				time.Sleep(DB_PING_TIMEOUT_SECS * time.Second)
+			}
+		}
+	}
+
+	return db
+}
+
+func ensureDatabaseConnection(settings *model.SqlSettings, dataSource, connType string, lifecycleContext context.Context) *dbsql.DB {
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), DB_PING_TIMEOUT_SECS*time.Second)
+		defer cancel()
+
+		select {
+		case <-lifecycleContext.Done():
```

Comment:
I am wondering about the use of the lifecycleContext here. If the DB is not available, `app.NewServer` won't even start. And therefore, the signal handler code wouldn't even execute. So the context will never be cancelled, and therefore, this loop will actually never exit gracefully unless I am mistaken.

If you think about it, even the earlier version suffered from the same problem. If the DB was down, the server would be stuck pinging the DB, and if you send SIGINT, it wouldn't exit gracefully, but just abort.

At the same time, having an infinite loop in the code is not a great idea.

What are your thoughts on making the DB_PING_ATTEMPTS and pingInterval to be configurable? That way the code behaves as before in dev environments, and in production, the cloud team has full control over how long a DB can be expected to be down and set the ping_attempts and ping_interval appropriately. This also avoids the need for it to be a feature flag, as setting the old settings effectively makes it behave like the old code.

Reply:
aha, that is true since we are waiting for the os signals after we have the server successfully initialized on the `commands/server.go`. I guess we can work around this by passing the signals channel to the supplier, but thinking this even more it's starting to be a bit hairy.

Maybe if we step back and define what access semantics we would like from our storage layer, can help us shape what needed here.

My view is something like this: `Server` needs a `storage service` object. The storage service owns the storage connection lifecycle. So I guess that boils down to, avoid obtaining a connection when initializing the supplier? With that, we can "configure" the server with its required dependencies just fine. The app will lock on waiting for `os.Signal` and when a request comes and we will need something from storage, then we can either return an error(cannot obtain connection) which will result in a `500` or fulfill the request.

> What are your thoughts on making the DB_PING_ATTEMPTS and pingInterval to be configurable? That way the code behaves as before in dev environments, and in production, the cloud team has full control over how long a DB can be expected to be down and set the ping_attempts and ping_interval appropriately. This also avoids the need for it to be a feature flag, as setting the old settings effectively makes it behave like the old code.

Totally, that is the path of least resistance.

Reply:
I think this goes back to my comment [here](https://github.com/mattermost/mattermost-server/pull/16266#issuecomment-728281461). In short, the server needs a storage object _with an active connection_. That's the thing here. The DB _is_ a required dependency for the server to start.

Even if you ignore the schema migrations, there are other services which depend on the DB when the server starts.